src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertificatePal.cs

```diff
@@ -16,14 +18,74 @@ public static ICertificatePal FromHandle(IntPtr handle)
             return new OpenSslX509CertificateReader(Interop.libcrypto.X509_dup(handle));
         }
 
-        public static ICertificatePal FromBlob(byte[] rawData, string password, X509KeyStorageFlags keyStorageFlags)
+        public static unsafe ICertificatePal FromBlob(byte[] rawData, string password, X509KeyStorageFlags keyStorageFlags)
         {
-            return new OpenSslX509CertificateReader(rawData);
+            // If we can see a hyphen, assume it's PEM.  Otherwise try DER-X509, then fall back to DER-PKCS12.
+            SafeX509Handle cert;
+
+            // PEM
+            if (rawData[0] == '-')
+            {
+
+                using (SafeBioHandle bio = Interop.libcrypto.BIO_new(Interop.libcrypto.BIO_s_mem()))
+                {
+                    Interop.libcrypto.CheckValidOpenSslHandle(bio);
+
+                    Interop.libcrypto.BIO_write(bio, rawData, rawData.Length);
+                    cert = Interop.libcrypto.PEM_read_bio_X509_AUX(bio, IntPtr.Zero, IntPtr.Zero, IntPtr.Zero);
+                }
+
+                Interop.libcrypto.CheckValidOpenSslHandle(cert);
+            }
+
+            // DER-X509
+            cert = Interop.libcrypto.OpenSslD2I((ptr, b, i) => Interop.libcrypto.d2i_X509(ptr, b, i), rawData, checkHandle: false);
+
+            if (cert != null && !cert.IsInvalid)
```

Comment:
There are currently no cases where a null SafeHandle would be returned from the d2i helper.  I can get rid of all of the null halves of the check.